File: src/my_cpp_pkg/src/battery.cpp
#include "my_robot_interfaces/srv/battery_status.hpp"
#include "rclcpp/rclcpp.hpp"

class BatteryNode : public rclcpp::Node {
public:
    BatteryNode() : Node("battery_node") {
        batteryState_ = 100;
        timer_ = this->create_wall_timer(
            std::chrono::seconds(1), std::bind(&BatteryNode::setBatteryState, this));
        batteryClinet_ = this->create_client<my_robot_interfaces::srv::BatteryStatus>("set_led");
    };

    void setBatteryState() {
        if (batteryState_) {
            std::this_thread::sleep_for((std::chrono::seconds(6)));
            batteryState_ = false;
            threads_.push_back(std::thread(std::bind(&BatteryNode::callBatteryClient, this, batteryState_)));
            RCLCPP_INFO(this->get_logger(), "Battery Low");
        } else {
            std::this_thread::sleep_for((std::chrono::seconds(3)));
            batteryState_ = true;
            threads_.push_back(std::thread(std::bind(&BatteryNode::callBatteryClient, this, batteryState_)));
        }
    };

    void callBatteryClient(bool batteryState) {
        while (!batteryClinet_->wait_for_service(std::chrono::seconds(1)))
            RCLCPP_WARN(this->get_logger(), "Waiting for the server");

        auto req = std::make_shared<my_robot_interfaces::srv::BatteryStatus::Request>();

        if (batteryState) {
            req->led_number = 3;
            req->state = true;
        } else {
            req->led_number = 3;
            req->state = true;
        }

        auto future = batteryClinet_->async_send_request(req);

        try {
            auto res = future.get();
            RCLCPP_INFO(this->get_logger(), "response = %d", res->success);

        } catch (std::exception const &) {
            RCLCPP_ERROR(this->get_logger(), "Service call failed");
        };
    };

private:
    bool batteryState_;
    std::vector<std::thread> threads_;
    rclcpp::TimerBase::SharedPtr timer_;
    rclcpp::Client<my_robot_interfaces::srv::BatteryStatus>::SharedPtr batteryClinet_;
};

int main(int argc, char **argv) {
    rclcpp::init(argc, argv);
    auto node = std::make_shared<BatteryNode>();
    rclcpp::spin(node);
    rclcpp::shutdown();
};

File: src/my_py_pkg/my_py_pkg/battery.py
import time
import rclpy
from rclpy.node import Node
from my_robot_interfaces.srv import BatteryStatus


class BatteryNode(Node):
    def __init__(self):
        super().__init__("battery_node")
        self.battery_state__ = 0
        self.timer_ = self.create_timer(1, self.get_battery_state)

        self.battery_status_client = self.create_client(
            BatteryStatus, "set_led")

    def get_battery_state(self):
        if not self.battery_state__:
            time.sleep(3)
            self.battery_state__ = 100
            self.call_battery_status_client(battery_low=False)

        else:
            time.sleep(6)
            self.battery_state__ = 0
            self.call_battery_status_client(battery_low=True)

    def call_battery_status_client(self, battery_low: bool):
        while not self.battery_status_client.wait_for_service(1):
            self.get_logger().warn("Waiting for the clinet")

        req = BatteryStatus.Request()
        if battery_low:
            req.led_number = 3
            req.state = True
        else:
            req.led_number = 3
            req.state = False

        future = self.battery_status_client.call_async(req)
        future.add_done_callback(self.callback_battery_status_client)

    def callback_battery_status_client(self, future):
        try:
            response = future.result()
            self.get_logger().info("response = " + str(response.success))

        except Exception as e:
            self.get_logger().error("service call failed %r" % (e,))


def main(args=None):
    rclpy.init(args=args)
    node = BatteryNode()
    rclpy.spin(node)
    rclpy.shutdown()


if __name__ == "__main__":
    main()
